### DISCUSSION TITLE
Advice on testing code which uses `Ferrum::Browser`?

### DISCUSSION THREAD

**Discussion by postmodern** [rubycdp/ferrum#347] [Q&A]
Say I write a class which uses `Ferrum::Browser`. How do I then test that class, such that the tests will work in GitHub Actions? I see from Ferrum's own specs there's quite a bit of setup code in `spec/spec_helper.rb`, even a `Ferrum::Server` which is used to serve content up to `Ferrum::Browser`. I was wondering if there was boilerplate code I could use to setup RSpec tests. Could I use `file://` URLs instead of a local HTTP server?

I at least figured out that I need to call `browser-actions/setup-chrome@latest` in the GitHub Actions workflow., and call `Ferrum::Browser#quit` to cleanup the browser instance.

Thanks in advance.

**Comment by route** [comment]
I think you can try to serve a file or there's a https://github.com/rubycdp/ferrum#interceptoptions with `request.respond`, but I like the idea of making something to make ferrum easy to test in the other projects. So if I understand correctly you have a class that uses Ferrum to visit some urls and now we have to write specs for that class, but we also need to kinda stub network requests just for Ferrum, right?

**Comment by postmodern** [comment]
@route correct. I want to test scenarios where Ferrum will visit multiple URLs in order or visits a URL which causes multiple requests for JavaScript/CSS/etc. I could try to find websites that I could test against, but I would prefer to keep the tests self-contained and have the HTML contained within the tests. I suppose I could define a large `browser.on(:request) { ... }` rule with multiple `if request.match?(url1)` checks for each URL?

**Reply by route** [reply]
Correct, but that would be ugly, woudn't it?) but at least you could try and tell me if it works as expected, then we could use this approach for stubbing.

**Comment by dgmora** [comment]
Hello, I am in a similar position where I want to use Vessel to crawl some urls, but I'm unsure if there's a simple way to mock those requests in tests. Has there been any significant change in this area? Or is there some documentation / other gem I could take a look at? Thanks!

**Comment by route** [comment]
There was no change, but everything ferrum uses under spec can and should be extracted. One downside of it is that if we extract this code and put it to libs we would need to add extra dependencies of puma/webrick/rack + sinatra to a gem that in general doesn't require it.
